File: 2023/day-01/main.js
// https://adventofcode.com/2023/day/1

const fs = require("fs");

const input = fs
  .readFileSync(__dirname + "/input", { encoding: "utf8" })
  .split("\n");

const numberMap = {
  1: true,
  2: true,
  3: true,
  4: true,
  5: true,
  6: true,
  7: true,
  8: true,
  9: true,
};

const stringMap = {
  one: "1",
  two: "2",
  three: "3",
  four: "4",
  five: "5",
  six: "6",
  seven: "7",
  eight: "8",
  nine: "9",
};

function part1() {
  function findFromLeft(str) {
    for (let i = 0; i < str.length; i++) {
      const n = str[i];
      if (numberMap[n]) return n;
    }
  }

  function findFromRight(str) {
    for (let i = str.length; i >= 0; i--) {
      const n = str[i];
      if (numberMap[n]) return n;
    }
  }

  let answer = 0;

  for (const line of input) {
    const first = findFromLeft(line);
    const last = findFromRight(line);
    answer += Number(first + last);
  }

  return answer;
}

function part2() {
  function findFromLeft(str) {
    for (let i = 0; i < str.length; i++) {
      const n = str[i];
      if (numberMap[n]) return n;

      for (let j = 3; j <= 5; j++) {
        const seq = str.substring(i, i + j);
        const n = stringMap[seq];
        if (n) return n;
      }
    }
  }

  function findFromRight(str) {
    for (let i = str.length; i >= 0; i--) {
      const n = str[i];
      if (numberMap[n]) return n;

      for (let j = 3; j <= 5; j++) {
        const seq = str.substring(i - j, i);
        const n = stringMap[seq];
        if (n) return n;
      }
    }
  }

  let answer = 0;

  for (const line of input) {
    const first = findFromLeft(line);
    const last = findFromRight(line);
    answer += Number(first + last);
  }

  return answer;
}

console.log("2023 Day 01 - part 1:", part1());
console.log("2023 Day 01 - part 2:", part2());

File: 2023/day-01/main.c
#include <stdio.h>
#include <stdlib.h>
#include <string.h>

char str_to_digit(char* str) {
    int len = strlen(str);

    if (len == 3) {
        if (strcmp(str, "one") == 0) {
            return '1';
        } else if (strcmp(str, "two") == 0) {
            return '2';
        } else if (strcmp(str, "six") == 0) {
            return '6';
        }
    } else if (len == 4) {
        if (strcmp(str, "four") == 0) {
            return '4';
        } else if (strcmp(str, "five") == 0) {
            return '5';
        } else if (strcmp(str, "nine") == 0) {
            return '9';
        }
    } else {
        if (strcmp(str, "three") == 0) {
            return '3';
        } else if (strcmp(str, "seven") == 0) {
            return '7';
        } else if (strcmp(str, "eight") == 0) {
            return '8';
        }
    }

    return '\0';
}

int solution(FILE *file, int part) {
    rewind(file);

    int answer = 0;
    char buff[256];
    char str_num[3];
    str_num[2] = '\0'; // string termination

    while (fgets(buff, sizeof(buff), file)) {    
        str_num[0] = '\0'; // indicate not assigned

        for (int i = 0; buff[i] != '\0'; i++) {
            char c = buff[i];

            if (c >= '1' && c <= '9') {
                if (str_num[0] == '\0') {
                    str_num[0] = c;
                }

                str_num[1] = c;
            }

            if (part != 2) continue;

            for (int j = 3; j <= 5; j++) {
                char str_seq[j];

                for (int k = 0; k < j; k++) {
                    str_seq[k] = buff[i + k];
                }

                char d = str_to_digit(str_seq);

                if (d != '\0') {
                    if (str_num[0] == '\0') {
                        str_num[0] = d;
                    }

                    str_num[1] = d;
                }
            }
        }

        answer += atoi(str_num);
    }

    return answer;
}

int main() {
    FILE *file = fopen("input", "r");

    int part1 = solution(file, 1);
    int part2 = solution(file, 2);

    fclose(file);

    printf("part 1: %i\n", part1);
    printf("part 2: %i\n", part2);

    return 0;
}
